dis_snek/api/voice/audio.py

```diff
@@ -0,0 +1,251 @@
+import asyncio
+import audioop
+import subprocess  # noqa: S404
+import threading
+import time
+from abc import ABC, abstractmethod
+from pathlib import Path
+from typing import Union, Optional
+
+from _cffi_backend import buffer
+from yt_dlp import YoutubeDL
+
+__all__ = ["AudioBuffer", "BaseAudio", "Audio", "AudioVolume", "YTDLAudio"]
+
+ytdl = YoutubeDL(
+    {
+        "format": "bestaudio/best",
+        "outtmpl": "%(extractor)s-%(id)s-%(title)s.%(ext)s",
+        "restrictfilenames": True,
+        "noplaylist": True,
+        "nocheckcertificate": True,
+        "ignoreerrors": False,
+        "logtostderr": False,
+        "quiet": True,
+        "no_warnings": True,
+        "default_search": "auto",
+        "source_address": "0.0.0.0",  # noqa: S104
+    }
+)
+
+
+class AudioBuffer:
+    def __init__(self) -> None:
+        self._buffer = bytearray()
+        self._lock = threading.Lock()
+        self.initialised = threading.Event()
+
+    def __len__(self) -> int:
+        return len(self._buffer)
+
+    def extend(self, data: bytes) -> None:
+        """
+        Extend the buffer with additional data.
+
+        Args:
+            data: The data to add
+        """
+        with self._lock:
+            self._buffer.extend(data)
+
+    def read(self, total_bytes: int) -> bytearray:
+        """
+        Read `total_bytes` bytes of audio from the buffer.
+
+        Args:
+            total_bytes: Amount of bytes to read.
+
+        Returns:
+            Desired amount of bytes
+        """
+        with self._lock:
+            view = memoryview(self._buffer)
+            self._buffer = bytearray(view[total_bytes:])
+            data = bytearray(view[:total_bytes])
+            if 0 < len(data) < total_bytes:
+                # pad incomplete frames with 0's
+                data.extend(b"\0" * (total_bytes - len(data)))
+            return data
+
+
+class BaseAudio(ABC):
+    """Base structure of the audio."""
+
+    locked_stream: bool
+    """Prevents the audio task from closing automatically when no data is received."""
+    needs_encode: bool
+    """Does this audio data need encoding with opus?"""
+    bitrate: Optional[int]
+    """Optionally specify a specific bitrate to encode this audio data with"""
+
+    def __del__(self) -> None:
+        self.cleanup()
+
+    def cleanup(self) -> None:
+        """A method to optionally cleanup after this object is no longer required."""
+        ...
+
+    @property
+    def audio_complete(self) -> bool:
+        """A property to tell the player if more audio is expected."""
+        return False
+
+    @abstractmethod
+    def read(self, frame_size: int) -> bytes:
+        """
+        Reads frame_size ms of audio from source.
+
+        returns:
+            bytes of audio
+        """
+        ...
+
+
+class Audio(BaseAudio):
```

Comment:
Thats a niche use-case, that imo should be handled by the user themselves, not the library. Namely because we wont know anything about the bytestream.